node/building_blocks.re

```diff
@@ -58,6 +58,53 @@ let is_current_producer = (state, ~key) => {
 // TODO: bad naming
 // TODO: check if block must have published a new snapshot
 let is_signable = (state, block) => {
+  open Protocol.Operation.Side_chain;
+  let Node.{
+        trusted_validator_membership_change,
+        protocol: {last_seen_membership_change_timestamp, _},
+        _,
+      } = state;
+  let allowed_op_reception_delay = 0.200; // Assumed max network latency
+  let current_time = Unix.time();
+  let allowed_op_window = current_time +. allowed_op_reception_delay;
+  let next_allowed_membership_change_timestamp =
+    last_seen_membership_change_timestamp +. 24. *. 60. *. 60.;
+  let rec contains_only_trusted_add_validator_op = result_so_far =>
+    fun
+    | [] => result_so_far
+    | [h, ...t] =>
+      switch (h.kind) {
+      | Add_validator(validator, timestamp) =>
+        if (Trusted_validators_membership_change.(
+              Set.mem(
+                {address: validator.address, action: Add},
+                trusted_validator_membership_change,
+              )
+            )
+            && timestamp >= current_time
+            && timestamp <= allowed_op_window
+            && timestamp > next_allowed_membership_change_timestamp) {
+          contains_only_trusted_add_validator_op(result_so_far, t);
+        } else {
+          false;
+        }
+      | Remove_validator(validator, timestamp) =>
+        if (Trusted_validators_membership_change.(
+              Set.mem(
+                {address: validator.address, action: Remove},
+                trusted_validator_membership_change,
+              )
+            )
+            && timestamp >= current_time
+            && timestamp <= allowed_op_window
+            && timestamp > next_allowed_membership_change_timestamp) {
+          contains_only_trusted_add_validator_op(result_so_far, t);
```

Comment:
This logic is duplicated from the case above and it's a really  important piece of domain logic, please abstract it in some meaningful function

Reply:
Logic now simpler in the next commit now that timestamp is no longer  a part of operation and doesn't have to travel over the wire.